art/attacks/extraction/copycat_cnn.py

```diff
@@ -0,0 +1,151 @@
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+
+import numpy as np
+
+from art import NUMPY_DTYPE
+from art.attacks.attack import Attack
+from art.classifiers.classifier import Classifier
+from art.utils import to_categorical
+
+
+logger = logging.getLogger(__name__)
+
+
+class CopycatCNN(Attack):
+    """
+    Implementation of the copycat cnn attack from Jacson et al. (2018).
+
+    | Paper link: https://arxiv.org/abs/1806.05476
+    """
+    attack_params = Attack.attack_params + ['batch_size', 'nb_epochs', 'nb_stolen']
+
+    def __init__(self, classifier, batch_size=1, nb_epochs=10, nb_stolen=1):
+        """
+        Create a copycat cnn attack instance.
+
+        :param classifier: A victim classifier.
+        :type classifier: :class:`.Classifier`
+        :param batch_size: Size of batches.
+        :type batch_size: `int`
+        :param nb_epochs: Number of epochs to use for training.
+        :type nb_epochs: `int`
+        :param nb_stolen: Number of examples to be stolen.
+        :type nb_stolen: `int`
+        """
+        super(CopycatCNN, self).__init__(classifier=classifier)
+
+        params = {'batch_size': batch_size,
+                  'nb_epochs': nb_epochs,
+                  'nb_stolen': nb_stolen}
+        self.set_params(**params)
+
+    def generate(self, x, y=None, **kwargs):
+        """
+        Generate a thieved classifier.
+
+        :param x: An array with the source input to the victim classifier.
+        :type x: `np.ndarray`
+        :param y: Target values (class labels) one-hot-encoded of shape (nb_samples, nb_classes) or indices of shape
+                  (nb_samples,). Not used in this attack.
+        :type y: `np.ndarray` or `None`
+        :param thieved_classifier: A thieved classifier to be stolen.
+        :type thieved_classifier: :class:`.Classifier`
+        :return: The stolen classifier.
+        :rtype: :class:`.Classifier`
+        """
+        # Warning to users if y is not None
+        if y is not None:
+            logger.warning("This attack does not use the provided label y.")
+
+        # Check the size of the source input vs nb_stolen
+        if x.shape[0] < self.nb_stolen:
+            logger.warning("The size of the source input is smaller than the number of expected stolen examples.")
+
+        # Check if there is a thieved classifier provided for training
+        thieved_classifier = kwargs.get('thieved_classifier')
+        if thieved_classifier is None or not isinstance(thieved_classifier, Classifier):
+            raise ValueError('A thieved classifier is needed.')
+
+        # Select data to attack
+        selected_x = self._select_data(x)
+
+        # Query the victim classifier
+        fake_labels = self._query_label(selected_x)
+
+        # Train the thieved classifier
+        thieved_classifier.fit(x=selected_x, y=fake_labels, batch_size=self.batch_size, nb_epochs=self.nb_epochs)
+
```

Comment:
Does it make sense to ensure `thieved_classifier` is a `Classifier` or just that it has a `fit` function?

Reply:
As discussed with @beat-buesser , we would like to keep the thieved_classifier an instance of the ART Classifier for 2 reasons: 1. If we just check for the availability of the fit function, we may not know the signature of that function for use, 2. ART supports a BlackClassifier that allows any type of classifier to be wrapped as an ART classifier.